From a6ebc694dcde640e5af43c5b079bdd30e2af9c5e Mon Sep 17 00:00:00 2001
From: Edgar Cherkasov <edgar.cherkasov@snt.rs>
Date: Fri, 9 Sep 2022 11:05:39 +0200
Subject: [PATCH 1/1] tegra_machine_driver: add support of RT5640 audio codec

Cherry picked cd5936dbd1bd56e0db0e92a0244352b3daebbb53 from
https://AIM-Linux@dev.azure.com/AIM-Linux/EPC-R7200/_git/kernel
repository and adapted to 5.10 kernel.
---
 .../machine_drivers/tegra_machine_driver.c    | 52 ++++++++++++++++++-
 1 file changed, 51 insertions(+), 1 deletion(-)

diff --git a/sound/soc/tegra-alt/machine_drivers/tegra_machine_driver.c b/sound/soc/tegra-alt/machine_drivers/tegra_machine_driver.c
index 28dd5f479..c50c34d72 100644
--- a/sound/soc/tegra-alt/machine_drivers/tegra_machine_driver.c
+++ b/sound/soc/tegra-alt/machine_drivers/tegra_machine_driver.c
@@ -29,6 +29,7 @@
 #include <sound/soc.h>
 #include <dt-bindings/sound/tas2552.h>
 #include "rt5659.h"
+#include "rt5640.h"
 #include "sgtl5000.h"
 #include "tegra_asoc_machine_alt.h"
 #include "tegra210_xbar_alt.h"
@@ -181,6 +182,16 @@ static int tegra_machine_dai_init(struct snd_soc_pcm_runtime *runtime,
 		}
 	}
 
+	rtd = snd_soc_get_pcm_runtime(card, "rt5640-playback");
+	if (rtd) {
+		err = snd_soc_dai_set_sysclk(rtd->codec_dai, RT5640_SCLK_S_MCLK,
+					     aud_mclk, SND_SOC_CLOCK_IN);
+		if (err < 0) {
+			dev_err(card->dev, "codec_dai clock not set\n");
+			return err;
+		}
+	}
+
 	rtd = snd_soc_get_pcm_runtime(card, "rt565x-playback");
 	if (rtd) {
 		err = snd_soc_dai_set_sysclk(rtd->codec_dai, RT5659_SCLK_S_MCLK,
@@ -392,6 +403,42 @@ static int tegra_machine_fepi_init(struct snd_soc_pcm_runtime *rtd)
 	return 0;
 }
 
+static int tegra_machine_rt5640_init(struct snd_soc_pcm_runtime *rtd)
+{
+	struct snd_soc_card *card = rtd->card;
+	struct snd_soc_jack *jack;
+	int err;
+
+	jack = devm_kzalloc(card->dev, sizeof(struct snd_soc_jack), GFP_KERNEL);
+	if (!jack)
+		return -ENOMEM;
+
+	err = snd_soc_card_jack_new(card, "Headset Jack", SND_JACK_HEADSET,
+				    jack, NULL, 0);
+	if (err) {
+		dev_err(card->dev, "Headset Jack creation failed %d\n", err);
+		return err;
+	}
+
+	err = tegra_machine_add_codec_jack_control(card, rtd, jack);
+	if (err) {
+		dev_err(card->dev, "Failed to add jack control: %d\n", err);
+		return err;
+	}
+
+	/* single button supporting play/pause */
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_0, KEY_MEDIA);
+
+	/* multiple buttons supporting play/pause and volume up/down */
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_1, KEY_MEDIA);
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_2, KEY_VOLUMEUP);
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_3, KEY_VOLUMEDOWN);
+
+	snd_soc_dapm_sync(&card->dapm);
+
+	return 0;
+}
+
 static int tegra_machine_rt565x_init(struct snd_soc_pcm_runtime *rtd)
 {
 	struct snd_soc_card *card = rtd->card;
@@ -446,7 +493,10 @@ static int codec_init(struct tegra_machine *machine)
 		if (!dai_links[i].name)
 			continue;
 
-		if (strstr(dai_links[i].name, "rt565x-playback") ||
+		if (strstr(dai_links[i].name, "rt5640-playback") ||
+		    strstr(dai_links[i].name, "rt5640-codec-sysclk-bclk1"))
+			dai_links[i].init = tegra_machine_rt5640_init;
+		else if (strstr(dai_links[i].name, "rt565x-playback") ||
 		    strstr(dai_links[i].name, "rt565x-codec-sysclk-bclk1"))
 			dai_links[i].init = tegra_machine_rt565x_init;
 		else if (strstr(dai_links[i].name, "fe-pi-audio-z-v2"))
